core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -361,37 +466,51 @@ public ReserveResult tryReserve(PendingAcquire acquire)
                         }
                         return false;
                     })
-                    .max(comparing(node -> nodesRemainingMemory.get(node.getNodeIdentifier())));
+                    .collect(toImmutableList());
 
-            if (selectedNode.isEmpty()) {
+            if (candidates.isEmpty()) {
                 return ReserveResult.NONE_MATCHING;
             }
 
-            boolean selectedCandidateMatches = false;
-            if (nodesRemainingMemory.get(selectedNode.get().getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.get())) {
+            InternalNode selectedNode = candidates.stream()
+                    .max(comparing(node -> nodesRemainingMemoryRuntimeAdjusted.get(node.getNodeIdentifier())))
+                    .orElseThrow();
+
+            if (nodesRemainingMemoryRuntimeAdjusted.get(selectedNode.getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.getNodeIdentifier())) {
                 // there is enough unreserved memory on the node
                 // OR
                 // there is not enough memory available on the node but the node is empty so we cannot to better anyway
 
                 // todo: currant logic does not handle heterogenous clusters best. There is a chance that there is a larger node in the cluster but
                 //       with less memory available right now, hence that one was not selected as a candidate.
-                selectedCandidateMatches = true;
+                // mark memory reservation
+                subtractFromRemainingMemory(selectedNode.getNodeIdentifier(), acquire.getMemoryLease());
+                return ReserveResult.reserved(selectedNode);
             }
 
-            // mark memory reservation
-            nodesRemainingMemory.compute(
-                    selectedNode.get().getNodeIdentifier(),
-                    (key, free) -> free - acquire.getMemoryLease());
+            // if selected node cannot be used right now, select best one ignoring runtime memory usage and reserve space ther
```

Comment:
nit: maybe explain it why it is needed

Reply:
MAde it:
```
            // If selected node cannot be used right now, select best one ignoring runtime memory usage and reserve space there
            // for later use. This is important from algorithm liveliness perspective. If we did not reserve space for a task which
            // is too big to be scheduled right now, it could be starved by smaller tasks coming later.

```